src/main/java/org/opentripplanner/routing/trippattern/TripTimes.java

```diff
@@ -474,6 +478,66 @@ public boolean tripAcceptable(final State state0, final int stopIndex) {
         return true;
     }
 
+    /**
+     * Returns true if the current state chain contains a banned trip, or banned trip/stop index or banned trip sequence
+     */
+    public boolean tripOrTripSequenceIsBanned(State state0, final int stopIndex) {
+        final RoutingRequest options = state0.getOptions();
+
+        // check if an exact trip ID is banned
+        final BannedStopSet banned = options.bannedTrips.get(trip.getId());
+        if (banned != null && banned.contains(stopIndex)) {
+            return true;
+        }
+
+        // check if a trip ID sequence is banned, but first return false if there are not banned sequences
+        if (options.bannedTripSequences.size() == 0) {
+            return false;
+        }
+
+        // compile list of trips seen so far in states (use ArrayList constructor to ensure a mutable list is returned)
+        List<FeedScopedId> tripsInState = new ArrayList<>(Arrays.asList(trip.getId()));
+        State tripFindingState = state0;
+        Trip lastTrip = trip;
+        while (tripFindingState != null) {
+            if (tripFindingState.getBackEdge() != null) {
+                Trip trip = tripFindingState.getBackTrip();
+                if (trip != null && trip != lastTrip) {
+                    tripsInState.add(trip.getId());
+                    lastTrip = trip;
+                }
+            }
+            tripFindingState = tripFindingState.getBackState();
+        }
+
+        // reverse trips if finding for depart-at search to make sure the list is in chronological order
+        if (!options.arriveBy) {
+            Collections.reverse(tripsInState);
+        }
+
+        // Check for the trip state sequence in all banned trip sequences. This will return true even for sequences that
+        // have additional trips included between a banned sequence. (ex: if trip A > trip B is banned, trip A >
+        // Trip C > Trip B will be flagged as a banned sequence).
+        for (List<FeedScopedId> bannedTripSequence : options.bannedTripSequences) {
+            int tripsInStateIdx = 0;
+            int bannedSequenceIdx = 0;
+            while (tripsInStateIdx < tripsInState.size() && bannedSequenceIdx < bannedTripSequence.size()) {
+                if (tripsInState.get(tripsInStateIdx).equals(bannedTripSequence.get(bannedSequenceIdx))) {
+                    // trip ID matches, also advance index of banned sequence
+                    bannedSequenceIdx++;
+                }
+                tripsInStateIdx++;
+            }
+            if (bannedSequenceIdx == bannedTripSequence.size()) {
+                // a matching trip sequence has been found!
+                return true;
+            }
```

Comment:
This block could become a `for` loop (using an index if it needs to be explicit):
```java
for (FeedScopeId tripInState : tripsInState) {
  if (tripInState.equals(bannedTripSequence.get(bannedSequenceIdx)) {
    bannedSequenceIdx++;
    if (bannedSequenceIdx == bannedTripSequence.size()) {
      // A matching trip sequence has been found!
      return true;
    }
  }
}
```

Reply:
This does seem more readable somehow and is fewer lines of code. I refactored to this in 6adec27faa2e744f94cd91e4ae4d337a5c25c7c7.